### ISSUE TITLE
Misleading Error with &dyn

### ISSUE THREAD

**Issue by C0D3-M4513R** [rust-lang/rust#95598]
[Given the following code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6d2a1542dd043198badb4121790bf504)

```rust
use std::path::Path;
// use std::ptr::{null, null_mut};

///Returns a function, which compares a &str against a name
fn cmp(name:&dyn ToString)
{}
fn cmp1(name:&dyn AsRef<Path>)
{}
fn main(){
    cmp("test");
    cmp1("test");
}
```

The current output is:

```
error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> src/main.rs:10:9
   |
10 |     cmp("test");
   |     --- ^^^^^^ doesn't have a size known at compile-time
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sized` is not implemented for `str`
   = note: required for the cast to the object type `dyn ToString`

error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> src/main.rs:11:10
   |
11 |     cmp1("test");
   |     ---- ^^^^^^ doesn't have a size known at compile-time
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sized` is not implemented for `str`
   = note: required for the cast to the object type `dyn AsRef<Path>`

For more information about this error, try `rustc --explain E0277`.
```

Stating, that &str does not have a known size at compile time is just wrong.
I think that rust sees, that the function parameter has a Borrowed type and that the function call also provides a borrowed type. So rust probably procedes to check the actual types 'behind' the borrow.

It is just, that here I would expect a message asking me if I would have liked to borrow here, since that would solve this error.

**Comment by C0D3-M4513R** [comment]
My use case for this is the following function, which returns a predicate.
I have the borrow infront of the dyn, to make the type Sized (so it can be used in the closure).
```rust
fn cmp<P:AsRef<Path>>(name:P) -> impl Fn(&dyn AsRef<Path>)->bool
{
    move |s| {
        return name.as_ref().ends_with(s)
    }
}
```
To call it, I need to do the following.
The first borrow should be obvious. The second is due to `AsRef`.
```rust
cmp(&&"test")
```

Please correct me, but I do not think, that I can write this any better right now (without inlining the function). I

**Comment by SNCPlay42** [comment]
>Stating, that &str does not have a known size at compile time is just wrong.

The message does not state that `&str` does not implement `Sized`, it states that `str` doesn't, which is true. And it does say why it needs this ("required for the cast to the object type `dyn ToString`"): only types of known size can be coerced into trait objects. This is because the representation of `&dyn Trait` provides no way to discern the length of the `str` being pointed to.

**Comment by C0D3-M4513R** [comment]
> > Stating, that &str does not have a known size at compile time is just wrong.
> 
> The message does not state that `&str` does not implement `Sized`, it states that `str` doesn't, which is true. And it does say why it needs this ("required for the cast to the object type `dyn ToString`"): only types of known size can be coerced into trait objects. This is because the representation of `&dyn Trait` provides no way to discern the length of the `str` being pointed to.

Yes, but the whole message is missing the point, of what is actually happening. The Message is very unclear in what is says. Take the error message, without the hints. Misses completely what is going on.
```
  --> src/main.rs:10:9
   |
10 |     cmp("test");
   |     --- ^^^^^^ doesn't have a size known at compile-time
   |     |
   |     required by a bound introduced by this call
```

Also can you please explain to me, why `str` needs to be Sized eventhough AsRef specifically lifts this requirement?
